qiskit/circuit/parameterexpression.py

```diff
@@ -36,12 +38,14 @@ def __init__(self, symbol_map, expr):
             expr (sympy.Expr): Expression of sympy.Symbols.
         """
         self._parameter_symbols = symbol_map
+        self._parameters = set(self._parameter_symbols)
         self._symbol_expr = expr
+        self._names = {p.name: p for p in self._parameters}
```

Comment:
I ran the parameterized circuit building and binding benchmarks and compared this with lazy initializing `self._names` and it didn't really show any changes:
```
Benchmarks that have stayed the same:

       before           after         ratio
     [b4cd85fb]       [4b5c27e0]
     <tune-parameterexpression~1>       <tune-parameterexpression>
              n/a              n/a      n/a  circuit_construction.ParameterizedCircuitBindBench.time_bind_params(20, 128, 131072)
              n/a              n/a      n/a  circuit_construction.ParameterizedCircuitBindBench.time_bind_params(20, 128, 2048)
              n/a              n/a      n/a  circuit_construction.ParameterizedCircuitBindBench.time_bind_params(20, 128, 32768)
              n/a              n/a      n/a  circuit_construction.ParameterizedCircuitBindBench.time_bind_params(20, 128, 8192)
              n/a              n/a      n/a  circuit_construction.ParameterizedCircuitBindBench.time_bind_params(20, 2048, 131072)
              n/a              n/a      n/a  circuit_construction.ParameterizedCircuitBindBench.time_bind_params(20, 2048, 32768)
              n/a              n/a      n/a  circuit_construction.ParameterizedCircuitBindBench.time_bind_params(20, 2048, 8192)
              n/a              n/a      n/a  circuit_construction.ParameterizedCircuitBindBench.time_bind_params(20, 32768, 131072)
              n/a              n/a      n/a  circuit_construction.ParameterizedCircuitBindBench.time_bind_params(20, 8, 128)
              n/a              n/a      n/a  circuit_construction.ParameterizedCircuitBindBench.time_bind_params(20, 8, 131072)
              n/a              n/a      n/a  circuit_construction.ParameterizedCircuitBindBench.time_bind_params(20, 8, 2048)
              n/a              n/a      n/a  circuit_construction.ParameterizedCircuitBindBench.time_bind_params(20, 8, 32768)
              n/a              n/a      n/a  circuit_construction.ParameterizedCircuitBindBench.time_bind_params(20, 8, 8192)
              n/a              n/a      n/a  circuit_construction.ParameterizedCircuitBindBench.time_bind_params(20, 8192, 131072)
              n/a              n/a      n/a  circuit_construction.ParameterizedCircuitBindBench.time_bind_params(20, 8192, 32768)
              n/a              n/a      n/a  circuit_construction.ParameterizedCircuitConstructionBench.time_build_parameterized_circuit(20, 128, 131072)
              n/a              n/a      n/a  circuit_construction.ParameterizedCircuitConstructionBench.time_build_parameterized_circuit(20, 128, 2048)
              n/a              n/a      n/a  circuit_construction.ParameterizedCircuitConstructionBench.time_build_parameterized_circuit(20, 128, 32768)
              n/a              n/a      n/a  circuit_construction.ParameterizedCircuitConstructionBench.time_build_parameterized_circuit(20, 128, 8192)
              n/a              n/a      n/a  circuit_construction.ParameterizedCircuitConstructionBench.time_build_parameterized_circuit(20, 2048, 131072)
              n/a              n/a      n/a  circuit_construction.ParameterizedCircuitConstructionBench.time_build_parameterized_circuit(20, 2048, 32768)
              n/a              n/a      n/a  circuit_construction.ParameterizedCircuitConstructionBench.time_build_parameterized_circuit(20, 2048, 8192)
              n/a              n/a      n/a  circuit_construction.ParameterizedCircuitConstructionBench.time_build_parameterized_circuit(20, 32768, 131072)
              n/a              n/a      n/a  circuit_construction.ParameterizedCircuitConstructionBench.time_build_parameterized_circuit(20, 8, 128)
              n/a              n/a      n/a  circuit_construction.ParameterizedCircuitConstructionBench.time_build_parameterized_circuit(20, 8, 131072)
              n/a              n/a      n/a  circuit_construction.ParameterizedCircuitConstructionBench.time_build_parameterized_circuit(20, 8, 2048)
              n/a              n/a      n/a  circuit_construction.ParameterizedCircuitConstructionBench.time_build_parameterized_circuit(20, 8, 32768)
              n/a              n/a      n/a  circuit_construction.ParameterizedCircuitConstructionBench.time_build_parameterized_circuit(20, 8, 8192)
              n/a              n/a      n/a  circuit_construction.ParameterizedCircuitConstructionBench.time_build_parameterized_circuit(20, 8192, 131072)
              n/a              n/a      n/a  circuit_construction.ParameterizedCircuitConstructionBench.time_build_parameterized_circuit(20, 8192, 32768)
       1.67±0.01s       1.79±0.02s     1.07  circuit_construction.ParameterizedCircuitBindBench.time_bind_params(20, 32768, 128)
       2.08±0.01m       2.19±0.02m     1.05  circuit_construction.ParameterizedCircuitConstructionBench.time_build_parameterized_circuit(20, 131072, 8)
         660±40ms         691±40ms     1.05  circuit_construction.ParameterizedCircuitConstructionBench.time_build_parameterized_circuit(20, 32768, 2048)
       1.71±0.02s       1.80±0.08s     1.05  circuit_construction.ParameterizedCircuitBindBench.time_bind_params(20, 32768, 2048)
       1.82±0.03s       1.88±0.05s     1.03  circuit_construction.ParameterizedCircuitBindBench.time_bind_params(20, 32768, 8192)
      1.96±0.04ms      2.01±0.04ms     1.03  circuit_construction.ParameterizedCircuitBindBench.time_bind_params(20, 8, 8)
        7.75±0.2s        7.95±0.1s     1.03  circuit_construction.ParameterizedCircuitBindBench.time_bind_params(20, 131072, 32768)
         721±30ms         736±40ms     1.02  circuit_construction.ParameterizedCircuitConstructionBench.time_build_parameterized_circuit(20, 32768, 8192)
        12.3±0.1s       12.5±0.05s     1.02  circuit_construction.ParameterizedCircuitConstructionBench.time_build_parameterized_circuit(20, 131072, 128)
       6.85±0.01s       6.95±0.07s     1.01  circuit_construction.ParameterizedCircuitBindBench.time_bind_params(20, 131072, 128)
       1.01±0.02s       1.02±0.09s     1.01  circuit_construction.ParameterizedCircuitConstructionBench.time_build_parameterized_circuit(20, 32768, 32768)
         452±10ms         455±10ms     1.01  circuit_construction.ParameterizedCircuitBindBench.time_bind_params(20, 8192, 2048)
        7.18±0.1s       7.23±0.08s     1.01  circuit_construction.ParameterizedCircuitBindBench.time_bind_params(20, 131072, 8192)
       7.59±0.4ms      7.63±0.02ms     1.01  circuit_construction.ParameterizedCircuitBindBench.time_bind_params(20, 128, 8)
       1.71±0.02s       1.72±0.02s     1.00  circuit_construction.ParameterizedCircuitBindBench.time_bind_params(20, 32768, 8)
       6.71±0.02s       6.73±0.01s     1.00  circuit_construction.ParameterizedCircuitBindBench.time_bind_params(20, 131072, 2048)
       3.16±0.02s       3.16±0.05s     1.00  circuit_construction.ParameterizedCircuitConstructionBench.time_build_parameterized_circuit(20, 131072, 2048)
       2.26±0.09s        2.25±0.1s     0.99  circuit_construction.ParameterizedCircuitBindBench.time_bind_params(20, 32768, 32768)
          548±3ms          545±4ms     0.99  circuit_construction.ParameterizedCircuitBindBench.time_bind_params(20, 8192, 8192)
          411±6ms          408±3ms     0.99  circuit_construction.ParameterizedCircuitBindBench.time_bind_params(20, 8192, 8)
        6.84±0.1s       6.78±0.03s     0.99  circuit_construction.ParameterizedCircuitBindBench.time_bind_params(20, 131072, 8)
          101±1ms         99.4±3ms     0.99  circuit_construction.ParameterizedCircuitBindBench.time_bind_params(20, 2048, 8)
       4.04±0.03s       3.97±0.02s     0.98  circuit_construction.ParameterizedCircuitConstructionBench.time_build_parameterized_circuit(20, 131072, 131072)
          102±1ms        101±0.9ms     0.98  circuit_construction.ParameterizedCircuitBindBench.time_bind_params(20, 2048, 128)
       46.1±0.2ms       45.4±0.2ms     0.98  circuit_construction.ParameterizedCircuitConstructionBench.time_build_parameterized_circuit(20, 2048, 8)
        9.07±0.1s       8.85±0.05s     0.98  circuit_construction.ParameterizedCircuitBindBench.time_bind_params(20, 131072, 131072)
       2.76±0.02s        2.69±0.1s     0.97  circuit_construction.ParameterizedCircuitConstructionBench.time_build_parameterized_circuit(20, 131072, 8192)
         829±70ms         806±70ms     0.97  circuit_construction.ParameterizedCircuitConstructionBench.time_build_parameterized_circuit(20, 32768, 128)
          416±4ms          403±5ms     0.97  circuit_construction.ParameterizedCircuitBindBench.time_bind_params(20, 8192, 128)
        3.38±0.2s        3.27±0.2s     0.97  circuit_construction.ParameterizedCircuitConstructionBench.time_build_parameterized_circuit(20, 32768, 8)
       2.87±0.03s       2.78±0.05s     0.97  circuit_construction.ParameterizedCircuitConstructionBench.time_build_parameterized_circuit(20, 131072, 32768)
         866±30μs         836±30μs     0.97  circuit_construction.ParameterizedCircuitConstructionBench.time_build_parameterized_circuit(20, 8, 8)
          137±3ms        132±0.6ms     0.96  circuit_construction.ParameterizedCircuitBindBench.time_bind_params(20, 2048, 2048)
          302±6ms          289±9ms     0.96  circuit_construction.ParameterizedCircuitConstructionBench.time_build_parameterized_circuit(20, 8192, 8)
          165±4ms          158±2ms     0.95  circuit_construction.ParameterizedCircuitConstructionBench.time_build_parameterized_circuit(20, 8192, 128)
       3.91±0.2ms      3.74±0.05ms     0.95  circuit_construction.ParameterizedCircuitConstructionBench.time_build_parameterized_circuit(20, 128, 128)
          182±5ms          171±2ms     0.94  circuit_construction.ParameterizedCircuitConstructionBench.time_build_parameterized_circuit(20, 8192, 2048)
       3.09±0.1ms      2.90±0.04ms     0.94  circuit_construction.ParameterizedCircuitConstructionBench.time_build_parameterized_circuit(20, 128, 8)
         8.79±1ms      8.16±0.05ms     0.93  circuit_construction.ParameterizedCircuitBindBench.time_bind_params(20, 128, 128)
          260±9ms          239±1ms     0.92  circuit_construction.ParameterizedCircuitConstructionBench.time_build_parameterized_circuit(20, 8192, 8192)
         40.9±5ms       37.4±0.7ms     0.91  circuit_construction.ParameterizedCircuitConstructionBench.time_build_parameterized_circuit(20, 2048, 128)
         63.8±3ms       57.6±0.1ms    ~0.90  circuit_construction.ParameterizedCircuitConstructionBench.time_build_parameterized_circuit(20, 2048, 2048)

BENCHMARKS NOT SIGNIFICANTLY CHANGED.
```

But, I'll make the change since it can't hurt

Reply:
Done in: https://github.com/Qiskit/qiskit-terra/pull/5028/commits/4b5c27e0e0dc7a5a019477f5c1d95a74654fc634